### ISSUE TITLE
Destructuring non-iterables causes potential unsoundness -- does not raise errors

### ISSUE THREAD

**Issue by niieani** [facebook/flow#4990]
I was bitten by an error after misremembering the signature for `forEach` on a `Map`. I tried to destructure value & key, but only discovered the bug after running it. Interestingly, TypeScript correctly discovers the error.

```js
const someMap : Map<string, {}> = new Map()

// incorrect usage, but no error:
someMap.forEach(([value, key]) => {
  console.log(value)
})

// correct usage:
someMap.forEach((value, key) => {
  console.log(value)
})
```

- [Flow Try](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAxnAdgZwC5g5wC2ApgLICGADmAFxhXUA8+ATgJZYDmANGAG8AvgD4wAXjBZSSJgAoAlKlREyTAHRQ4bAKKV0ACzlyA2gDdKMAK6l+Aa1IBPALoKJYgajBhMuODFJ1eG45C2tSJSElFRIKGk1tPUNjMJt7JzdxDy8fbCIAoLgQ1IjUKNQgA)
- [TypeScript Playground](https://www.typescriptlang.org/play/#src=const%20someMap%20%3A%20Map%3Cstring%2C%20%7B%7D%3E%20%3D%20new%20Map()%0D%0A%0D%0AsomeMap.forEach((%5Bvalue%2C%20key%5D)%20%3D%3E%20%7B%0D%0A%20%20console.log(value)%0D%0A%7D)%0D%0A%0D%0AsomeMap.forEach((value%2C%20key)%20%3D%3E%20%7B%0D%0A%20%20console.log(value)%0D%0A%7D)%0D%0A)

**Comment by minedeljkovic** [comment]
Possibly related:
#1512
#2612

**Comment by StevenLangbroek** [comment]
That's because given a string key, this is valid:

```js
const [a, b] = 'ab';
console.log(a); // 'a'
console.log(b); // 'b'
```

**Comment by niieani** [comment]
@StevenLangbroek no, `value` is an object, and that's the first parameter of the `forEach` callback. It's `key` that is a `string`. The bug is real.

**Comment by StevenLangbroek** [comment]
Sorry, you're absolutely right, didn't realise it flipped.

**Comment by niieani** [comment]
Actually, a simpler way to reproduce is:

```js
// @flow
const [a, b] = {} // should error, but doesn't
```

https://flow.org/try/#0PTAEAEDMBsHsHcBQBjWA7AzgF1AbQIYA0oARgLqgC8oA3gL6JA

This is a runtime error when run in V8, so definitely a bug:
```
TypeError: {} is not iterable
```

**Comment by gkz** [comment]
Fixed in cfe00503857c1e3dc2566f8d56c3cf522c017b36
